### PR TITLE
Prevent repartitioning of certain operator's direct children (#1731)

### LINKED ISSUE #1731: Holistic Repartition Optimisation
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The `Repartition` optimizer recurses into plans and looks for operators with an output partition count less than the target partition count, and on finding such a partition adds a `RepartitionExec` node.

Unfortunately this does not take into account how this might impact the concurrency of the plan as a whole.

For example, in IOx it is fairly common to have a `UnionExec` with a list of many `IOxReadFilterNode` as children. Each of these `IOxReadFilterNode` is a single partition, and therefore the `Repartition` optimizer repartitions the output of each of these `IOxReadFilterNode`.

This leads to a partition explosion, with the `UnionExec` going from potentially having `x` partitions, to `n * x` where `n` is the target parallelism. In some of our plans this is leading to thousands of partitions, which is not ideal :grin: 

A similar quirk can occur when there is an operator further up the chain that imposes some partitioning constraint, e.g. a sort, but isn't the direct parent. I'm fairly certain that I've seen plans generated that `RepartitionExec` into a `LimitExec` into a `CoalescePartitionsExec`. 

**Describe the solution you'd like**

I don't really know how to solve this, but I wanted to flag up that the current greedy approach can be pretty sub-optimal, and that perhaps a more holistic optimizer that can take into account the plan as a whole might be beneficial

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #1731.

 # Rationale for this change

Not all operators benefit from additional concurrency, in particular those that are not CPU-intensive (e.g. `UnionExec`, `ProjectionExec`) or are not embarrassingly parallel (e.g. `LimitExec`). Introducing additional partitioning is unlikely to yield a performance benefit for these operators, and may in fact make execution slower. It also makes the plans harder to follow.

# What changes are included in this PR?

Adds a new trait method `ExecutionPlan::should_repartition_children` that allows operators to indicate that their direct children should not be repartitioned.

# Are there any user-facing changes?

There is a new optional `ExecutionPlan` method

### CODE REVIEW

File: datafusion/src/physical_plan/projection.rs [hunk 1]
```diff
@@ -185,6 +185,10 @@ impl ExecutionPlan for ProjectionExec {
             self.expr.iter().map(|(e, _)| Arc::clone(e)),
         )
     }
+
+    fn should_repartition_children(&self) -> bool {
+        false
```
Comment (Dandandan) [inline datafusion/src/physical_plan/projection.rs:190]: A processing heavy projection (e.g. parsing during cast, complex case/when statement, etc.) benefits from having multiple partitions, to run in parallel.

File: datafusion/src/physical_plan/repartition.rs [hunk 2]
```diff
@@ -259,6 +259,10 @@ impl ExecutionPlan for RepartitionExec {
     fn statistics(&self) -> Statistics {
         self.input.statistics()
     }
+
+    fn should_repartition_children(&self) -> bool {
+        false
```
Comment (Dandandan) [inline datafusion/src/physical_plan/repartition.rs:264]: Hash-repartition is quite expensive, so might benefit from being repartitioned.

**Review by Dandandan — COMMENTED** [review]
We

**Comment by Dandandan** [comment]
I put some comments on ProjectionExec and RepartitionExec which might in some cases benefit from being repartitioned (maybe not in many cases, but local round-robin repartitioning is AFAIK very cheap compared to e.g. hashing string columns or parsing numbers etc.)

**Comment by tustvold** [comment]
@Dandandan I've reverted the changes to those, the main operator I'm interested in excluding is UnionExec as it is showing up in IOx scan plans and yielding 3000 odd partitions as it repartitions every parquet file 28 times. :laughing: 

For context one of the reasons I'm interested in reducing the concurrency is somewhat stupid, the plans and the resulting metrics and tracing become HUGE, but I also can't help but feel that always pursuing the maximum concurrency is liable to hit a wall somewhere - probably memory :grin:

**Comment by alamb** [comment]
> For context one of the reasons I'm interested in reducing the concurrency is somewhat stupid, the plans and the resulting metrics and tracing become HUGE, but I also can't help but feel that always pursuing the maximum concurrency is liable to hit a wall somewhere - probably memory 😁


I think pursuing the maximum concurrency specified in `target_partitions` is a good strategy, but it needs to be complimented by a more sophisticated scheduler, that breaks the plans down into smaller pieces, perhaps as described / explained in  https://github.com/apache/arrow-datafusion/issues/64

File: datafusion/src/physical_optimizer/repartition.rs [hunk 3]
```diff
@@ -107,93 +105,148 @@ impl PhysicalOptimizerRule for Repartition {
...
+        let optimizer = Repartition {};
+
+        let optimized = optimizer.optimize(
+            hash_aggregate(filter_exec(parquet_exec())),
+            &ExecutionConfig::new().with_target_partitions(10),
         )?;
 
+        let plan = displayable(optimized.as_ref()).indent().to_string();
+
+        let expected = &[
```
Comment (alamb) [inline datafusion/src/physical_optimizer/repartition.rs:213]: I think this is a good improvement to the tests 👍

File: datafusion/src/physical_plan/mod.rs [hunk 4]
```diff
@@ -135,14 +135,32 @@ pub trait ExecutionPlan: Debug + Send + Sync {
...
+
     /// Specifies the data distribution requirements of all the children for this operator
     fn required_child_distribution(&self) -> Distribution {
         Distribution::UnspecifiedDistribution
     }
+
+    /// Returns `true` if the direct children of this `ExecutionPlan` should be repartitioned
+    /// to introduce greater concurrency to the plan
+    ///
+    /// The default implementation returns `true` unless `Self::request_child_distribution`
```
Comment (alamb) [inline datafusion/src/physical_plan/mod.rs:153]:
```suggestion
    /// The default implementation returns `true` unless `Self::required_child_distribution`
```

File: datafusion/src/physical_plan/limit.rs [hunk 5]
```diff
@@ -300,6 +300,10 @@ impl ExecutionPlan for LocalLimitExec {
             _ => Statistics::default(),
         }
     }
+
+    fn should_repartition_children(&self) -> bool {
```
Comment (alamb) [inline datafusion/src/physical_plan/limit.rs:304]:
```suggestion
    /// No reason to repartition children as this node is just limiting each input partition. 
    fn should_repartition_children(&self) -> bool {
```

File: datafusion/src/physical_optimizer/repartition.rs [hunk 6]
```diff
@@ -107,93 +105,148 @@ impl PhysicalOptimizerRule for Repartition {
...
-            .is_some());
-
+        let plan = displayable(optimized.as_ref()).indent().to_string();
+
+        let expected = &[
+            "HashAggregateExec: mode=Final, gby=[], aggr=[]",
+            "HashAggregateExec: mode=Partial, gby=[], aggr=[]",
+            "RepartitionExec: partitioning=RoundRobinBatch(10)",
+            "GlobalLimitExec: limit=100",
+            "LocalLimitExec: limit=100",
```
Comment (alamb) [inline datafusion/src/physical_optimizer/repartition.rs:241]: maybe worth a comment on the rationale here. Something like
```suggestion
            // Expect no repartition to happen for local limit
            "LocalLimitExec: limit=100",
```

File: datafusion/src/physical_optimizer/repartition.rs [hunk 7]
```diff
@@ -107,93 +105,148 @@ impl PhysicalOptimizerRule for Repartition {
...
+            "FilterExec: c1@0",
+            "RepartitionExec: partitioning=RoundRobinBatch(10)",
+            "GlobalLimitExec: limit=100",
+            "LocalLimitExec: limit=100", // Should not repartition for LocalLimitExec
+            "ParquetExec: limit=None, partitions=[x]",
+        ];
+
```
Comment (alamb) [inline datafusion/src/physical_optimizer/repartition.rs:248]: Could we also add a test for `UnionExec` as well? Mostly to guard against regression?

I notice that there is already coverage for repartitioning projection which is good 👍 
https://github.com/apache/arrow-datafusion/pull/1732/commits/c4c0a1ce876bcaff5096e19c44e5cc3ea0f38760

**Comment by Dandandan** [comment]
> > For context one of the reasons I'm interested in reducing the concurrency is somewhat stupid, the plans and the resulting metrics and tracing become HUGE, but I also can't help but feel that always pursuing the maximum concurrency is liable to hit a wall somewhere - probably memory grin
> 
> I think pursuing the maximum concurrency specified in `target_partitions` is a good strategy, but it needs to be complimented by a more sophisticated scheduler, that breaks the plans down into smaller pieces, perhaps as described / explained in #64

Agreed with this approach :100:

**Review by alamb — APPROVED** [review]
LGTM -- any final thoughts @Dandandan ?

**Review by Dandandan — APPROVED** [review]
Looks good!

**Comment by Dandandan** [comment]
Thanks @tustvold !

File: datafusion/src/physical_plan/limit.rs [hunk 8]
```diff
@@ -300,6 +300,11 @@ impl ExecutionPlan for LocalLimitExec {
...
+        // No reason to repartition children as this node is just limiting each input partition.
+        false
```
Comment (alamb) [inline datafusion/src/physical_plan/limit.rs:306]: This also fixed https://github.com/apache/arrow-datafusion/issues/423 I believe

Comment (tustvold) [reply datafusion/src/physical_plan/limit.rs:306]: As currently expressed this only prevent repartitioning of direct children, I think you need something more than that for sortedness as discussed on #424

**Comment by alamb** [comment]
Removing API change label as it was showing up prominently on the changelog and is not a breaking API change (really)

**Pull request merged**